drivers/serial/uart_numaker.c

```diff
@@ -0,0 +1,478 @@
+	UART_T *uart = config->uart;
+
+	UART_DisableInt(uart, UART_INTEN_RDAIEN_Msk);
+}
+
+static int uart_numaker_irq_rx_ready(const struct device *dev)
+{
+	const struct uart_numaker_config *config = dev->config;
+	UART_T *uart = config->uart;
+
+	return ((!UART_GET_RX_EMPTY(uart)) && (uart->INTEN & UART_INTEN_RDAIEN_Msk));
+}
+
+static void uart_numaker_irq_err_enable(const struct device *dev)
+{
+	const struct uart_numaker_config *config = dev->config;
+	UART_T *uart = config->uart;
+
+	UART_EnableInt(uart, UART_INTEN_BUFERRIEN_Msk | UART_INTEN_SWBEIEN_Msk);
+}
+
+static void uart_numaker_irq_err_disable(const struct device *dev)
+{
+	const struct uart_numaker_config *config = dev->config;
+	UART_T *uart = config->uart;
+
+	UART_DisableInt(uart, UART_INTEN_BUFERRIEN_Msk | UART_INTEN_SWBEIEN_Msk);
+}
+
+static int uart_numaker_irq_is_pending(const struct device *dev)
+{
+
+	return (uart_numaker_irq_tx_ready(dev) || (uart_numaker_irq_rx_ready(dev)));
+}
+
+static int uart_numaker_irq_update(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* nothing to be done */
+	return 1;
+}
+
+static void uart_numaker_irq_callback_set(const struct device *dev,
+					  uart_irq_callback_user_data_t cb, void *cb_data)
+{
+	struct uart_numaker_data *pData = dev->data;
+
+	pData->user_cb = cb;
+	pData->user_data = cb_data;
+}
+
+static void uart_numaker_isr(const struct device *dev)
+{
+	struct uart_numaker_data *pData = dev->data;
+
+	if (pData->user_cb) {
+		pData->user_cb(dev, pData->user_data);
+	}
+}
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static const struct uart_driver_api uart_numaker_driver_api = {
+	.poll_in = uart_numaker_poll_in,
+	.poll_out = uart_numaker_poll_out,
+	.err_check = uart_numaker_err_check,
+#ifdef CONFIG_UART_USE_RUNTIME_CONFIGURE
+	.configure = uart_numaker_configure,
+	.config_get = uart_numaker_config_get,
+#endif
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_numaker_fifo_fill,
+	.fifo_read = uart_numaker_fifo_read,
+	.irq_tx_enable = uart_numaker_irq_tx_enable,
+	.irq_tx_disable = uart_numaker_irq_tx_disable,
+	.irq_tx_ready = uart_numaker_irq_tx_ready,
+	.irq_tx_complete = uart_numaker_irq_tx_complete,
+	.irq_rx_enable = uart_numaker_irq_rx_enable,
+	.irq_rx_disable = uart_numaker_irq_rx_disable,
+	.irq_rx_ready = uart_numaker_irq_rx_ready,
+	.irq_err_enable = uart_numaker_irq_err_enable,
+	.irq_err_disable = uart_numaker_irq_err_disable,
+	.irq_is_pending = uart_numaker_irq_is_pending,
+	.irq_update = uart_numaker_irq_update,
+	.irq_callback_set = uart_numaker_irq_callback_set,
+#endif
+};
+
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_SCC
+#define CLOCK_CTRL_INIT(n) .clk_dev = DEVICE_DT_GET(DT_PARENT(DT_INST_CLOCKS_CTLR(n))),
+#else
+#define CLOCK_CTRL_INIT(n)
+#endif
+
+#ifdef CONFIG_PINCTRL
+#define PINCTRL_DEFINE(n) PINCTRL_DT_INST_DEFINE(n);
+#define PINCTRL_INIT(n)	  .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),
+#else
+#define PINCTRL_DEFINE(n)
+#define PINCTRL_INIT(n)
+#endif
```

Comment:
please don't make this conditional, pinctrl is the only way now to configure pins (we need to do some cleanup on this)

Reply:
OK, regard pinctrl as the only way.